elara/elarautil.py

```diff
@@ -4,62 +4,108 @@
 
 This source code is licensed under the BSD-style license found in the LICENSE file in the root directory of this source tree.
 """
+from typing import Dict
 
-from cryptography.fernet import Fernet
-import json
+import msgpack
 import os
-from .exceptions import FileAccessError, FileKeyError
+from zlib import crc32
+
+from cryptography.fernet import Fernet
+
+from .exceptions import FileAccessError, FileKeyError, LoadChecksumError, LoadIncompatibleDB
 
 
 class Util:
     @staticmethod
-    def readJSON(obj):
-        try:
-            curr_db = json.load(open(obj.path, "rb"))
-        except Exception:
-            # print("Read JSON error. File might be encrypted. Run in secure mode.")
-            raise FileAccessError(
-                "Read JSON error. File might be encrypted. Run in secure mode with key path."
-            )
-        return curr_db
+    def check_mag(mag):
+        return mag == b"ELDB"
 
     @staticmethod
-    def storeJSON(obj):
-        try:
-            json.dump(obj.db, open(obj.path, "wt"), indent=4)
-        except Exception:
-            raise FileAccessError(
-                "Store JSON error. File might be encrypted. Run in secure mode with key path."
-            )
+    def check_encrypted(version):
+        # if msb of version number is set the db is encrypted
+        return (version & (1 << 15)) != 0
+
+    @staticmethod
+    def read_plain_db(obj) -> Dict:
+        with open(obj.path, "rb") as fctx:
+            if not Util.check_mag(fctx.read(4)):
+                raise FileAccessError("File magic number not known")
+            version = int.from_bytes(fctx.read(2), "little", signed=False)
+            # check for encryption before trying anything
+            if Util.check_encrypted(version):
+                raise FileAccessError("This file is encrypted, run in secure mode")
+            checksum = int.from_bytes(fctx.read(4), "little", signed=False)
```

Comment:
`fctx.read(4)` should return the first 4 bytes right?( similarly  `fctx.read(2)`)
I'm not sure how the `Util.check_mag(fctx.read(4))` exactly works, a little unclear with the syntax

Reply:
mag here refers to magic number i.e the bytes sequence equivalent to ELDB in ascii (thus read 4 bytes), it's easier if you imagine a caret and if you tell python to read n bytes it will read n bytes and advance the caret, so if I tell python
```python
bytes_4 = fctx.read(4)
bytes_2 = fctx.read(2)
```
I have now read the first 6 bytes of a file in a 4 bytes piece and a 2 bytes one that's after the original 4 bytes read.

Reply:
I hope I have done a good enough job of explaining it

Reply:
btw you can use a series of seek functions to move the caret inside the file(works with other buffer like objects python has too) I have not used those in this code but it's worthy of note

Reply:
Thanks, that clears it up :)